components/raftstore/src/store/fsm/peer.rs

```diff
@@ -602,13 +605,35 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
         }
     }
 
+    #[inline]
+    pub fn handle_raft_ready_apply(&mut self, ready: &mut Ready, invoke_ctx: &InvokeContext) {
+        self.fsm.early_apply = ready
+            .committed_entries
+            .as_ref()
+            .and_then(|e| e.last())
+            .map_or(false, |e| {
+                self.fsm.peer.can_early_apply(e.get_term(), e.get_index())
+            });
+        if !self.fsm.early_apply {
+            return;
+        }
+        self.fsm
+            .peer
+            .handle_raft_ready_apply(self.ctx, ready, invoke_ctx);
+    }
+
     pub fn post_raft_ready_append(&mut self, mut ready: Ready, invoke_ctx: InvokeContext) {
         let is_merging = self.fsm.peer.pending_merge_state.is_some();
+        if !self.fsm.early_apply {
+            self.fsm
+                .peer
+                .handle_raft_ready_apply(self.ctx, &mut ready, &invoke_ctx);
+        }
         let res = self
             .fsm
             .peer
             .post_raft_ready_append(self.ctx, &mut ready, invoke_ctx);
```

Comment:
Reversing the order of `post_raft_ready_append` and `handle_raft_ready_apply` may has problems. 
When this peer needs not to apply snapshot, the work in `post_raft_ready_append` is: 
1. set `raft_state` and `last_term`
2. send msg to others.
For now, I am not sure if it's ok to reverse the order because I am not familiar with the lease and read index logic.

But I think it's better to set `raft_state` and `last_term` before calling `handle_raft_ready_apply` and send msg after saving data to disk. And the applying snapshot logic should execute after saving data to disk.
In my opinion, the rules can be
1. the actions about just changing memory data can do before the data is really saved to disk. (The code executes in order so data will save to disk soon)
2. the other actions should do after the data is really saved to disk. Such as sending msg to other peers, applying snapshot.
I think this way can reduce the thinking burden. 
What do you think？

Reply:
> ...it's better to set raft_state and last_term before calling handle_raft_ready_apply ...

I don't see why need to update those states before calling apply.

> send msg after saving data to disk. And the applying snapshot logic should execute after saving data to disk.

The procedure is still the same after the patch.

> What do you think？

I agree the rule, and it's exact how the PR is implemented.